src/testFixtures/groovy/grails/plugin/geb/WebDriverContainerHolder.groovy

```diff
@@ -0,0 +1,137 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package grails.plugin.geb
+
+import geb.Browser
+import groovy.transform.EqualsAndHashCode
+import org.openqa.selenium.WebDriver
+import org.openqa.selenium.chrome.ChromeOptions
+import org.openqa.selenium.remote.RemoteWebDriver
+import org.spockframework.runtime.extension.IMethodInvocation
+import org.spockframework.runtime.model.SpecInfo
+import org.testcontainers.Testcontainers
+import org.testcontainers.containers.BrowserWebDriverContainer
+
+import java.time.Duration
+
+/**
+ * Responsible for initializing a {@link org.testcontainers.containers.BrowserWebDriverContainer BrowserWebDriverContainer}
+ * per the Spec's {@link grails.plugin.geb.ContainerGebConfiguration ContainerGebConfiguration}.  This class will try to
+ * reuse the same container if the configuration matches the current container.
+ *
+ * @author James Daugherty
+ * @since 5.0
+ */
+class WebDriverContainerHolder {
+    RecordingSettings recordingSettings
+    WebDriverContainerConfiguration configuration
+    BrowserWebDriverContainer current
+
+    WebDriverContainerHolder(RecordingSettings recordingSettings) {
+        this.recordingSettings = recordingSettings
+    }
+
+    boolean isInitialized() {
+        current != null
+    }
+
+    boolean stop() {
+        if (!current) {
+            return false
+        }
+        current.stop()
+        current = null
+        configuration = null
+        return true
+    }
+
+    boolean matchesCurrentContainerConfiguration(WebDriverContainerConfiguration specConfiguration) {
+        specConfiguration == configuration
+    }
+
+    private static int getPort(IMethodInvocation invocation) {
+        try {
+            return (int) invocation.instance.metaClass.getProperty(invocation.instance, 'serverPort')
+        } catch (Exception ignore) {
+            throw new IllegalStateException('Test class must be annotated with @Integration for serverPort to be injected')
+        }
+    }
+
+    boolean reinitialize(IMethodInvocation invocation) {
+        WebDriverContainerConfiguration specConfiguration = new WebDriverContainerConfiguration(getPort(invocation), invocation.getSpec())
+        if (matchesCurrentContainerConfiguration(specConfiguration)) {
+            return false
+        }
+
+        if (isInitialized()) {
+            stop()
+        }
+
+        configuration = specConfiguration
+        current = new BrowserWebDriverContainer()
+        Testcontainers.exposeHostPorts(configuration.port)
+        current.tap {
+            addExposedPort(configuration.port)
+            withAccessToHost(true)
+            start()
+        }
+        if (!isDefaultHostname()) {
```

Comment:
Use groovy property access `if (!defaultHostname)` for readability.
I had named this with a verb `hostnameChanged` so the code would read `if (hostnameChanged)`. With `if (!defaultHostname)` it is not clear if we are checking for null or boolean.